internal/tools/fetch_logs.go

```diff
@@ -0,0 +1,83 @@
+package tools
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/mark3labs/mcp-go/mcp"
+	"github.com/mark3labs/mcp-go/server"
+	"github.com/tektoncd/mcp-server/internal/params"
+	taskruninformer "github.com/tektoncd/pipeline/pkg/client/injection/informers/pipeline/v1/taskrun"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+)
+
+func fetchTaskRunLogs() server.ServerTool {
+	return server.ServerTool{
+		Tool: mcp.NewTool("fetch_taskrun_logs",
+			mcp.WithDescription("Fetch the logs for a given TaskRun"),
+			mcp.WithString("name", mcp.Required(),
+				mcp.Description("Name of the TaskRun to fetch logs for"),
+			),
+			mcp.WithString("namespace",
+				mcp.Description("Namespace where the TaskRun is located"),
+				mcp.DefaultString("default"),
+			),
+		),
+		Handler: handlerFetchTaskRunLogs,
+	}
+}
+
+func handlerFetchTaskRunLogs(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+	name := request.Params.Arguments["name"].(string)
+	namespace, err := params.Optional[string](request, "namespace")
+	if err != nil {
+		return mcp.NewToolResultErrorFromErr("namespace must be a string", err), nil
```

Comment:
What is the reason to return error in mcp.CallToolResult struct instead of returning it as the second param?

Reply:
just how things were done previously - maybe @vdemeester can comment but it seems to not be consistent in other projects. The docs show the below while an example returns an err.

```
        case "divide":
            if y == 0 {
                return mcp.NewToolResultError("cannot divide by zero"), nil
            }
```

Reply:
@ksimon1 yeah, the error doesn't necessarily "convey" the error to the user in the end. And in the initial version of mcp-go I used, at least, returning an error was actually causing issues. We may want to check nowadays if the library could help here (like automatically transform our error into mcp errors, but…